### ISSUE TITLE
TS doesn't raise errors if type of the returned value is not assignable to the function's type 

### ISSUE THREAD

**Issue by que-etc** [Microsoft/TypeScript#23533]
**TypeScript Version:** 2.9.0-dev.20180419

**Search Terms:** 
 - void type 
 - function type

**Code**

```ts
type Func = () => void;

// Following error should be raised:
// Type 'number' is not assignable to type 'void'.
const func: Func = () => {
    return 42;
};

type Func2 = () => string;

// Shows error as expected.
const func2: Func2 = () => {
    return 42
}

// Shows error as expected.
const func3 = (value: string): void => {
    return 42;
};
```

**Expected behavior:**
If inferred type of the returned value (number) is not assignable to the function's type (void), following error should be raised:
`Type 'number' is not assignable to type 'void'.`

**Actual behavior:**
No error is shown.

**Playground Link:** [link](http://www.typescriptlang.org/play/index.html#src=type%20Func%20%3D%20()%20%3D%3E%20void%3B%0D%0A%0D%0A%2F%2F%20Following%20error%20should%20be%20raised%3A%0D%0A%2F%2F%20Type%20'number'%20is%20not%20assignable%20to%20type%20'void'.%0D%0Aconst%20func%3A%20Func%20%3D%20()%20%3D%3E%20%7B%0D%0A%20%20%20%20return%2042%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20Func2%20%3D%20()%20%3D%3E%20string%3B%0D%0A%0D%0A%2F%2F%20Shows%20error%20as%20expected.%0D%0Aconst%20func2%3A%20Func2%20%3D%20()%20%3D%3E%20%7B%0D%0A%20%20%20%20return%2042%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Shows%20error%20as%20expected.%0D%0Aconst%20func3%20%3D%20(value%3A%20string)%3A%20void%20%3D%3E%20%7B%0D%0A%20%20%20%20return%2042%3B%0D%0A%7D%3B%0D%0A)

**Comment by RyanCavanaugh** [comment]
Is there something we can do to make the FAQ more usable? The first hit for "void" is exactly the behavior you're describing. This is also listed on in the section titled  "Common 'bugs' that aren't bugs". https://github.com/Microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs

It seems like the documentation isn't doing its job.

**Comment by que-etc** [comment]
@RyanCavanaugh but if an inferred non-empty type has a higher priority, why TS doesn't use in subsequent calls:

```ts
type Func = () => void;

const func: Func = () => {
    return 42;
};

// Property 'toFixed' does not exist on type 'void'.
func().toFixed();
```

**Comment by RyanCavanaugh** [comment]
@que-etc Why *wouldn't* it?

**Comment by RyanCavanaugh** [comment]
If you wrote this code:
```ts
const a: Animal = fish;
// ...
a.swim();
```
I would consider it a type system defect to not error the call to `swim` - if I had *wanted* to have access to the subtype methods of the initializer, I wouldn't have written the type annotation in the first place.

**Comment by que-etc** [comment]
I'd expect TS to do so, because otherwise we have a discrepancy between the actually returned type and the declared one.

Still can't wrap my head around why the former type substitution is not being applied in this example:

``` ts
// Tells that 42 is not assignable to type 'void'
const func = (): void => {
    return 42;
};
```

Anyway, I'm sorry. From now on, I solemnly promise to double-check before submitting an issue like this one.

**Comment by RyanCavanaugh** [comment]
Returning a value from a `void`-annotated function is very likely indicative of a programmer error; this is *not* the case when using a non-`void`-returning function alias/expression in a place where only a `void` return type is required.

**Comment by que-etc** [comment]
> Bug report: 2 * 1 = 2 and 2 / 1 = 2, so 2 + 1 should also be 2.

Thanks :)

**Comment by kpdonn** [comment]
@RyanCavanaugh FWIW I think there's a pretty good argument for an error in the specific case where a freshly implemented function returns a value in a location that specifies a `void` return return type. `const func: Func = () => 11` indicates programmer error exactly as much as `const func = (): void => 22` does because there's no way for the `11` that is returned to ever be used by typed code without a type assertion.

I completely understand why it's a valid assignment, but I think it's essentially the same idea as excess property checks on fresh object literals. Having excess properties normally doesn't violate the assignability rules of TypeScript but brand new objects are handled differently because it almost always indicates a programmer mistake or at least misunderstanding since the excess properties would be unusable anywhere without a later type assertion.

**Comment by RyanCavanaugh** [comment]
@kpdonn is this code wrong?

```ts
var someOtherArr = [];
[1, 2, 3].forEach(el => someOtherArr.push(el));
```

**Comment by kpdonn** [comment]
Good point that's obviously totally valid although really only because of the implicit return behavior of arrow functions. If it was written with old style functions in this way:

```ts
var someOtherArr = [];
[1, 2, 3].forEach(function(el){
    return someOtherArr.push(el);
});
```
I'd say yes it's wrong or at least unintentional.

But I concede the point, arrow functions make it so that you very well might end up returning a value implicitly while writing perfectly legitimate code, and I don't think it's nearly an important enough issue to ask typescript to handle implicit returns from arrow functions different from explicit returns.

**Comment by RyanCavanaugh** [comment]
I agree that code is sketchy AF. I might see what happens if we break it.

It's a good example of how consistency is a false virtue. `x => y` "should be" identical to `function(x) { return y; }` but it's really easy to write code where one looks clearly OK and the other looks clearly wrong.

**Comment by RyanCavanaugh** [comment]
Interestingly we do have the notion of a side-effect free expression, so it is potentially possible to ban `return 42;` in that declaration without banning `return arr.push(el);` if it seems like "Return now along with doing this side effect" is a desirable pattern to some people
